File: finanze/infrastructure/repository/earnings_expenses/pending_flow_repository.py
from typing import Optional
from uuid import UUID, uuid4

from application.ports.pending_flow_port import PendingFlowPort
from domain.dezimal import Dezimal
from domain.earnings_expenses import FlowStatus, FlowType, PendingFlow
from infrastructure.repository.db.client import DBClient
from infrastructure.repository.earnings_expenses.queries import PendingFlowsQueries


def _map_row_to_pending_flow(row) -> PendingFlow:
    return PendingFlow(
        id=UUID(row["id"]),
        name=row["name"],
        amount=Dezimal(row["amount"]),
        currency=row["currency"],
        flow_type=FlowType(row["flow_type"]),
        category=row["category"],
        status=FlowStatus(row["status"]),
        date=row["date"],
        icon=row["icon"],
        status_changed_at=row["status_changed_at"],
    )


class PendingFlowRepository(PendingFlowPort):
    def __init__(self, client: DBClient):
        self._db_client = client

    async def save(self, flow: PendingFlow) -> PendingFlow:
        if flow.id is None:
            flow.id = uuid4()

        async with self._db_client.tx() as cursor:
            await cursor.execute(
                PendingFlowsQueries.INSERT,
                (
                    str(flow.id),
                    flow.name,
                    str(flow.amount),
                    flow.currency,
                    flow.flow_type.value,
                    flow.category,
                    flow.status.value,
                    flow.status_changed_at.isoformat()
                    if flow.status_changed_at
                    else None,
                    flow.date,
                    flow.icon,
                ),
            )
        return flow

    async def update(self, flow: PendingFlow):
        async with self._db_client.tx() as cursor:
            await cursor.execute(
                PendingFlowsQueries.UPDATE,
                (
                    flow.name,
                    str(flow.amount),
                    flow.currency,
                    flow.flow_type.value,
                    flow.category,
                    flow.status.value,
                    flow.status_changed_at.isoformat()
                    if flow.status_changed_at
                    else None,
                    flow.date,
                    flow.icon,
                    str(flow.id),
                ),
            )

    async def delete(self, flow_id: UUID):
        async with self._db_client.tx() as cursor:
            await cursor.execute(PendingFlowsQueries.DELETE_BY_ID, (str(flow_id),))

    async def get_all(self) -> list[PendingFlow]:
        async with self._db_client.read() as cursor:
            await cursor.execute(PendingFlowsQueries.GET_ALL)
            return [_map_row_to_pending_flow(row) for row in await cursor.fetchall()]

    async def get_by_id(self, flow_id: UUID) -> Optional[PendingFlow]:
        async with self._db_client.read() as cursor:
            await cursor.execute(PendingFlowsQueries.GET_BY_ID, (str(flow_id),))
            row = await cursor.fetchone()
            if row is None:
                return None
            return _map_row_to_pending_flow(row)

File: mobile/src/services/database/repositories/pendingFlowRepository.ts
import { DataManager } from "../dataManager"
import { PendingFlow, FlowType } from "@/domain"
import { parseDezimalValue } from "@/domain"
import { PendingFlowPort } from "@/application/ports"
import { PendingFlowsQueries } from "./queries"

function uuidV4(): string {
  // Prefer native implementation if available
  const anyCrypto = globalThis as any
  if (anyCrypto?.crypto?.randomUUID) {
    return anyCrypto.crypto.randomUUID()
  }

  // Fallback RFC4122-ish v4
  const bytes = Array.from({ length: 16 }, () =>
    Math.floor(Math.random() * 256),
  )
  bytes[6] = (bytes[6] & 0x0f) | 0x40
  bytes[8] = (bytes[8] & 0x3f) | 0x80
  const hex = bytes.map(b => b.toString(16).padStart(2, "0")).join("")
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`
}

export class PendingFlowRepository implements PendingFlowPort {
  constructor(private client: DataManager) {}

  private async tableExists(name: string): Promise<boolean> {
    const res = await this.client.query<{ name: string }>(
      "SELECT name FROM sqlite_master WHERE type='table' AND name = ?",
      [name],
    )
    return res.rows.length > 0
  }

  async save(flows: PendingFlow[]): Promise<void> {
    const hasTable = await this.tableExists("pending_flows")
    if (!hasTable) return

    await this.client.transaction(async () => {
      for (const flow of flows) {
        const id = flow.id ?? uuidV4()
        await this.client.execute(PendingFlowsQueries.INSERT, [
          id,
          flow.name,
          String(flow.amount),
          flow.currency,
          flow.flowType,
          flow.category ?? null,
          flow.enabled ? 1 : 0,
          flow.date ?? null,
          flow.icon ?? null,
        ])
        flow.id = id
      }
    })
  }

  async deleteAll(): Promise<void> {
    const hasTable = await this.tableExists("pending_flows")
    if (!hasTable) return
    await this.client.execute(PendingFlowsQueries.DELETE_ALL)
  }

  async getAll(): Promise<PendingFlow[]> {
    const hasTable = await this.tableExists("pending_flows")
    if (!hasTable) return []

    const res = await this.client.query<any>(PendingFlowsQueries.GET_ALL)
    return res.rows.map((row: any) => ({
      id: row.id,
      name: row.name,
      amount: parseDezimalValue(row.amount),
      currency: row.currency,
      flowType: (row.flow_type as FlowType) || FlowType.EXPENSE,
      category: row.category ?? null,
      enabled: !!row.enabled,
      date: row.date ?? null,
      icon: row.icon ?? null,
    }))
  }
}
